src/track/beatutils.cpp

```diff
@@ -5,44 +5,350 @@
+    }
+    auto tempoMedianFilter = MovingMedian(numberOfBeatsInFilteringWindow);
+    auto mostFrequentTempo = MovingMode(numberOfBeatsInFilteringWindow);
+    int currentBeat = -1;
+    int lastBeatChange = 0;
+    QMap<int, double> stableTemposByPosition;
+    // We start at beat 0 with the median bpm as rough guess of the "correct" tempo
+    stableTemposByPosition[lastBeatChange] = medianBpm;
+    // Here we are going to track the significant tempo changes over the track to make
+    // regions of stable tempos and indendently make const tempo or ironed grid for them
+    for (double tempo : tempoList) {
+        currentBeat += 1;
+        double filteredTempo = tempoMedianFilter.pushAndEvaluate(tempo);
+        double newStableTempo = mostFrequentTempo.pushAndEvaluate(filteredTempo);
+        // The analyzer has some jitter that causes a steady beat to fluctuate around the correct
+        // value so we don't consider changes to a neighboor value at the ordered tempo table
+        if (newStableTempo == stableTemposByPosition.last()) {
+            continue;
+            // Here we check if the new tempo is the right neighboor of the previous tempo
+        } else if (stableTemposByPosition.last() != frequencyOfTempos.lastKey() and
+                newStableTempo == (frequencyOfTempos.find(stableTemposByPosition.last()) + 1).key()) {
+            continue;
+            // Here we check if the new tempo is the left neighboor of the previous tempo
+        } else if (stableTemposByPosition.last() != frequencyOfTempos.firstKey() and
+                newStableTempo == (frequencyOfTempos.find(stableTemposByPosition.last()) - 1).key()) {
+            continue;
+        } else {
+            // This may not be case when our median window is even, we can't use it
+            // Because find() will return an iterator pointing to end that we will *
+            if (frequencyOfTempos.contains(newStableTempo)) {
+                lastBeatChange = currentBeat - tempoMedianFilter.lag() - mostFrequentTempo.lag();
+                // Since we used the median as guess of the first tempo
+                // we change position at 0 if we detect a change in very beginning 
+                if (lastBeatChange > kBeatsToCountTempo) {
+                    stableTemposByPosition[lastBeatChange] = newStableTempo;
+                } else {
+                    stableTemposByPosition[0] = newStableTempo;
+                }
+            }
+        }
+    }
+    // We also add the median as rough guess as our last tempo
+    // This way we always have both sentinels for the whole track
+    // as a constant tempo region if no significant tempo changes
+    stableTemposByPosition[tempoList.count() - 1] = medianBpm;
+    qDebug() << stableTemposByPosition;
+    return stableTemposByPosition;
+}
+
+void BeatUtils::removeSmallArrhythmic(
+        QVector<double>& beats, int sampleRate, QMap<int, double>& stableTemposByPosition) {
```

Comment:
IIRC we have a special sample rate type in `src/audio/types.h` that we should use here.

Reply:
I think we need to get rid here if the sample rate dependency. At this stage everything is sampled by the 0,01133 ms raster we use initialy.
Using BPM as intermediate values turns integer int doubles, which are harder to handle in the statistics filter. See my comments about bin size parameter there.